File: qc_checks.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');

// ── Extract the JSON from the HTML ──────────────────────────────────────────
const html = fs.readFileSync(
  '/Users/stefanhiekin/projects/fl_ag_campaign_site/campaign_schedule_outreach.html',
  'utf8'
);

const match = html.match(/window\.FL_AG_CAMPAIGN_DATA\s*=\s*(\[[\s\S]*?\]);\s*\n/);
if (!match) {
  // Try a greedier match if the first fails
  const m2 = html.match(/window\.FL_AG_CAMPAIGN_DATA\s*=\s*(\[[\s\S]*\])\s*;/);
  if (!m2) { console.error('Could not find FL_AG_CAMPAIGN_DATA'); process.exit(1); }
  var jsonStr = m2[1];
} else {
  var jsonStr = match[1];
}

let data;
try {
  data = JSON.parse(jsonStr);
} catch(e) {
  console.error('JSON parse error:', e.message);
  process.exit(1);
}

console.log(`Loaded ${data.length} records.\n`);

// ── Helpers ──────────────────────────────────────────────────────────────────
const isEmpty = v => v === null || v === undefined || (typeof v === 'string' && v.trim() === '');
const name = r => r['Organization_or_Event_Name'] || '(no name)';

// ════════════════════════════════════════════════════════════════════════════
// CHECK 1 — Completeness Audit
// ════════════════════════════════════════════════════════════════════════════
console.log('═'.repeat(70));
console.log('CHECK 1 — COMPLETENESS AUDIT');
console.log('═'.repeat(70));

const REQUIRED_FIELDS = [
  'Region',
  'County',
  'Category',
  'Subcategory',
  'Organization_or_Event_Name',
  'Contact_Name',
  'Website',
  'Source_URL',
];

const check1Issues = [];
data.forEach((rec, idx) => {
  REQUIRED_FIELDS.forEach(field => {
    if (isEmpty(rec[field])) {
      check1Issues.push({ idx, org: name(rec), field });
    }
  });
});

if (check1Issues.length === 0) {
  console.log('✓ No missing required fields found.\n');
} else {
  console.log(`Found ${check1Issues.length} blank required field(s):\n`);
  // Group by field for cleaner output
  const byField = {};
  check1Issues.forEach(({ idx, org, field }) => {
    if (!byField[field]) byField[field] = [];
    byField[field].push({ idx, org });
  });
  Object.entries(byField).forEach(([field, records]) => {
    console.log(`  Field: ${field} — ${records.length} blank(s)`);
    records.forEach(({ idx, org }) => {
      console.log(`    [${idx}] ${org}`);
    });
  });
  console.log();
}

// Summary counts per field
console.log('Blank count per required field:');
REQUIRED_FIELDS.forEach(field => {
  const count = data.filter(r => isEmpty(r[field])).length;
  console.log(`  ${field.padEnd(35)} ${count}`);
});
console.log();

// ════════════════════════════════════════════════════════════════════════════
// CHECK 4 — Date Validation
// ════════════════════════════════════════════════════════════════════════════
console.log('═'.repeat(70));
console.log('CHECK 4 — DATE VALIDATION');
console.log('═'.repeat(70));

const MONTH_NAMES = ['January','February','March','April','May','June',
                     'July','August','September','October','November','December'];
const DATE_REGEX = /^([A-Za-z]+)\s+(\d{1,2}),\s+(\d{4})$/;

const tbdRecords = [];
const badFormat = [];
const beforeApril2026 = [];
const outsideWindow = [];
const monthCounts = {};

data.forEach((rec, idx) => {
  const d = rec['Meeting_or_Event_Date'];
  if (isEmpty(d)) {
    tbdRecords.push({ idx, org: name(rec) });
    return;
  }
  const m = d.trim().match(DATE_REGEX);
  if (!m) {
    badFormat.push({ idx, org: name(rec), value: d });
    return;
  }
  const [, monthStr, dayStr, yearStr] = m;
  const monthIdx = MONTH_NAMES.findIndex(mn => mn.toLowerCase() === monthStr.toLowerCase());
  const year = parseInt(yearStr, 10);
  const month = monthIdx + 1; // 1-based

  if (monthIdx === -1) {
    badFormat.push({ idx, org: name(rec), value: d, reason: 'unknown month name' });
    return;
  }

  // Track distribution
  const key = `${monthStr} ${year}`;
  monthCounts[key] = (monthCounts[key] || 0) + 1;

  // Check before April 2026
  if (year < 2026 || (year === 2026 && month < 4)) {
    beforeApril2026.push({ idx, org: name(rec), value: d });
  }

  // Check outside April–August 2026 window
  if (year !== 2026 || month < 4 || month > 8) {
    outsideWindow.push({ idx, org: name(rec), value: d });
  }
});

// Bad format
if (badFormat.length === 0) {
  console.log('✓ All dated records match "Month D, YYYY" format.\n');
} else {
  console.log(`Bad date format — ${badFormat.length} record(s):`);
  badFormat.forEach(({ idx, org, value }) => console.log(`  [${idx}] ${org} → "${value}"`));
  console.log();
}

// Before April 2026
if (beforeApril2026.length === 0) {
  console.log('✓ No dates before April 2026.\n');
} else {
  console.log(`Dates BEFORE April 2026 — ${beforeApril2026.length} record(s):`);
  beforeApril2026.forEach(({ idx, org, value }) => console.log(`  [${idx}] ${org} → "${value}"`));
  console.log();
}

// Outside April–August 2026
if (outsideWindow.length === 0) {
  console.log('✓ All dates fall within April–August 2026 window.\n');
} else {
  console.log(`Dates OUTSIDE April–August 2026 window — ${outsideWindow.length} record(s):`);
  outsideWindow.forEach(({ idx, org, value }) => console.log(`  [${idx}] ${org} → "${value}"`));
  console.log();
}

// Month distribution
console.log('Month distribution (records with a date):');
// Sort by year then month
const sortedMonths = Object.entries(monthCounts).sort(([a], [b]) => {
  const [am, ay] = a.split(' '); const [bm, by] = b.split(' ');
  const ai = MONTH_NAMES.indexOf(am); const bi = MONTH_NAMES.indexOf(bm);
  return parseInt(ay) - parseInt(by) || ai - bi;
});
sortedMonths.forEach(([k, v]) => console.log(`  ${k.padEnd(20)} ${v}`));
console.log();

// TBD records
console.log(`TBD records (no date) — ${tbdRecords.length} total:`);
if (tbdRecords.length > 0) {
  tbdRecords.forEach(({ idx, org }) => console.log(`  [${idx}] ${org}`));
}
console.log();

// ════════════════════════════════════════════════════════════════════════════
// CHECK 5 — County / Region Validation
// ════════════════════════════════════════════════════════════════════════════
console.log('═'.repeat(70));
console.log('CHECK 5 — COUNTY / REGION VALIDATION');
console.log('═'.repeat(70));

const OFFICIAL_COUNTIES = new Set([
  'Alachua','Baker','Bay','Bradford','Brevard','Broward','Calhoun','Charlotte',
  'Citrus','Clay','Collier','Columbia','DeSoto','Dixie','Duval','Escambia',
  'Flagler','Franklin','Gadsden','Gilchrist','Glades','Gulf','Hamilton','Hardee',
  'Hendry','Hernando','Highlands','Hillsborough','Holmes','Indian River','Jackson',
  'Jefferson','Lafayette','Lake','Lee','Leon','Levy','Liberty','Madison','Manatee',
  'Marion','Martin','Miami-Dade','Monroe','Nassau','Okaloosa','Okeechobee','Orange',
  'Osceola','Palm Beach','Pasco','Pinellas','Polk','Putnam','Santa Rosa','Sarasota',
  'Seminole','St. Johns','St. Lucie','Sumter','Suwannee','Taylor','Union','Volusia',
  'Wakulla','Walton','Washington'
]);

// Known region→county mappings to verify
const KNOWN_MAPPINGS = {
  'Miami-Dade':   'South Florida',
  'Duval':        'Northeast Florida',
  'Hillsborough': 'Central Florida',
  'Escambia':     'Northwest Florida',
  'Leon':         'North Central Florida',
  'Orange':       'Central Florida',
  'Polk':         'Central Florida',
  'Palm Beach':   'South Florida',
  'Broward':      'South Florida',
  'Lee':          'Southwest Florida',
  'Collier':      'Southwest Florida',
  'Volusia':      'Northeast Florida',
  'Brevard':      'Central Florida',
  'Pinellas':     'Central Florida',
};

// Collect unique counties and their region(s)
const countyRegionMap = {}; // county → Set<region>
const regionCountyMap = {}; // region → Set<county>

data.forEach((rec, idx) => {
  const county = rec['County'];
  const region = rec['Region'];
  if (!isEmpty(county)) {
    if (!countyRegionMap[county]) countyRegionMap[county] = new Set();
    countyRegionMap[county].add(region || '(blank)');
  }
  if (!isEmpty(region)) {
    if (!regionCountyMap[region]) regionCountyMap[region] = new Set();
    regionCountyMap[region].add(county || '(blank)');
  }
});

const uniqueCounties = Object.keys(countyRegionMap).sort();
console.log(`Unique counties in data (${uniqueCounties.length}):`);
console.log('  ' + uniqueCounties.join(', ') + '\n');

// Invalid counties
const invalidCounties = uniqueCounties.filter(c => !OFFICIAL_COUNTIES.has(c));
if (invalidCounties.length === 0) {
  console.log('✓ All county names match the official Florida county list.\n');
} else {
  console.log(`Invalid/unrecognised counties — ${invalidCounties.length}:`);
  invalidCounties.forEach(c => console.log(`  "${c}" → found in regions: ${[...countyRegionMap[c]].join(', ')}`));
  console.log();
}

// Missing counties (in official list but not in data)
const missingCounties = [...OFFICIAL_COUNTIES].filter(c => !countyRegionMap[c]).sort();
console.log(`Counties in official list but NOT in data (${missingCounties.length}):`);
if (missingCounties.length > 0) {
  console.log('  ' + missingCounties.join(', '));
}
console.log();

// Verify known mappings
console.log('Verifying known County → Region mappings:');
let mappingErrors = 0;
Object.entries(KNOWN_MAPPINGS).forEach(([county, expectedRegion]) => {
  const actualRegions = countyRegionMap[county];
  if (!actualRegions) {
    console.log(`  [NOT IN DATA] ${county} — expected: ${expectedRegion}`);
  } else {
    const regions = [...actualRegions];
    const ok = regions.every(r => r === expectedRegion);
    if (ok) {
      console.log(`  ✓ ${county.padEnd(16)} → ${expectedRegion}`);
    } else {
      console.log(`  ✗ ${county.padEnd(16)} → expected "${expectedRegion}", found: ${regions.join(', ')}`);
      mappingErrors++;
    }
  }
});
console.log(mappingErrors === 0 ? '\n✓ All checked mappings are correct.\n' : `\n${mappingErrors} mapping error(s) found.\n`);

// Complete region→county listing
console.log('Complete Region → County mapping in data:');
Object.entries(regionCountyMap).sort(([a],[b]) => a.localeCompare(b)).forEach(([region, counties]) => {
  const sorted = [...counties].sort();
  console.log(`\n  ${region} (${sorted.length} counties):`);
  console.log(`    ${sorted.join(', ')}`);
});
console.log();

// ════════════════════════════════════════════════════════════════════════════
// CHECK 7 — "Unverified" Label Check
// ════════════════════════════════════════════════════════════════════════════
console.log('═'.repeat(70));
console.log('CHECK 7 — "UNVERIFIED" LABEL CHECK');
console.log('═'.repeat(70));

// Literal "Unverified" string anywhere in any field
const literalUnverified = [];
data.forEach((rec, idx) => {
  const hits = Object.entries(rec)
    .filter(([, v]) => typeof v === 'string' && v.includes('Unverified'))
    .map(([k]) => k);
  if (hits.length > 0) {
    literalUnverified.push({ idx, org: name(rec), fields: hits });
  }
});

if (literalUnverified.length === 0) {
  console.log('✓ No records contain the literal string "Unverified".\n');
} else {
  console.log(`Records containing literal "Unverified" — ${literalUnverified.length}:`);
  literalUnverified.forEach(({ idx, org, fields }) =>
    console.log(`  [${idx}] ${org} (in fields: ${fields.join(', ')})`));
  console.log();
}

// Count empty Contact_Name, Website, Source_URL (these render as "Unverified" in UI)
const emptyContactName = data.filter(r => isEmpty(r['Contact_Name']));
const emptyWebsite = data.filter(r => isEmpty(r['Website']));
const emptySourceURL = data.filter(r => isEmpty(r['Source_URL']));

console.log('Empty field counts (render as "Unverified" in UI):');
console.log(`  Contact_Name  : ${emptyContactName.length}`);
console.log(`  Website       : ${emptyWebsite.length}`);
console.log(`  Source_URL    : ${emptySourceURL.length}`);
console.log();

if (emptyContactName.length > 0) {
  console.log(`  Contact_Name blanks — ${emptyContactName.length} records (goal: ZERO):`);
  emptyContactName.forEach((rec, i) => {
    const idx = data.indexOf(rec);
    console.log(`    [${idx}] ${name(rec)}`);
  });
  console.log();
} else {
  console.log('  ✓ Contact_Name: ZERO blanks — goal met.\n');
}

if (emptyWebsite.length > 0) {
  console.log(`  Website blanks — ${emptyWebsite.length} records:`);
  emptyWebsite.forEach(rec => {
    const idx = data.indexOf(rec);
    console.log(`    [${idx}] ${name(rec)}`);
  });
  console.log();
}

if (emptySourceURL.length > 0) {
  console.log(`  Source_URL blanks — ${emptySourceURL.length} records:`);
  emptySourceURL.forEach(rec => {
    const idx = data.indexOf(rec);
    console.log(`    [${idx}] ${name(rec)}`);
  });
  console.log();
}

console.log('═'.repeat(70));
console.log('QC COMPLETE');
console.log('═'.repeat(70));

File: qc_checks.py
#!/usr/bin/env python3
"""QC checks for FL_AG_CAMPAIGN_DATA in campaign_schedule_outreach.html"""

import re
import json
import sys
from collections import defaultdict

# ── Extract JSON ─────────────────────────────────────────────────────────────
html_path = '/Users/stefanhiekin/projects/fl_ag_campaign_site/campaign_schedule_outreach.html'
with open(html_path, 'r', encoding='utf-8') as f:
    html = f.read()

# Find the assignment — grab everything between the first [ and the matching ]
m = re.search(r'window\.FL_AG_CAMPAIGN_DATA\s*=\s*(\[)', html)
if not m:
    sys.exit('ERROR: could not find FL_AG_CAMPAIGN_DATA')

start = m.start(1)
depth = 0
end = start
for i, ch in enumerate(html[start:], start):
    if ch == '[':
        depth += 1
    elif ch == ']':
        depth -= 1
        if depth == 0:
            end = i + 1
            break

json_str = html[start:end]
data = json.loads(json_str)
print(f'Loaded {len(data)} records.\n')

# ── Helpers ──────────────────────────────────────────────────────────────────
def is_empty(v):
    return v is None or (isinstance(v, str) and v.strip() == '')

def rec_name(r):
    return r.get('Organization_or_Event_Name') or '(no name)'

SEP = '═' * 70

# ════════════════════════════════════════════════════════════════════════════
# CHECK 1 — Completeness Audit
# ════════════════════════════════════════════════════════════════════════════
print(SEP)
print('CHECK 1 — COMPLETENESS AUDIT')
print(SEP)

REQUIRED_FIELDS = [
    'Region',
    'County',
    'Category',
    'Subcategory',
    'Organization_or_Event_Name',
    'Contact_Name',
    'Website',
    'Source_URL',
]

issues = []
for idx, rec in enumerate(data):
    for field in REQUIRED_FIELDS:
        if is_empty(rec.get(field)):
            issues.append({'idx': idx, 'org': rec_name(rec), 'field': field})

if not issues:
    print('✓ No missing required fields found.\n')
else:
    print(f'Found {len(issues)} blank required field(s):\n')
    by_field = defaultdict(list)
    for item in issues:
        by_field[item['field']].append(item)
    for field, records in by_field.items():
        print(f'  Field: {field} — {len(records)} blank(s)')
        for r in records:
            print(f'    [{r["idx"]}] {r["org"]}')
    print()

print('Blank count per required field:')
for field in REQUIRED_FIELDS:
    count = sum(1 for r in data if is_empty(r.get(field)))
    print(f'  {field:<35} {count}')
print()

# ════════════════════════════════════════════════════════════════════════════
# CHECK 4 — Date Validation
# ════════════════════════════════════════════════════════════════════════════
print(SEP)
print('CHECK 4 — DATE VALIDATION')
print(SEP)

MONTH_NAMES = ['January','February','March','April','May','June',
               'July','August','September','October','November','December']
DATE_RE = re.compile(r'^([A-Za-z]+)\s+(\d{1,2}),\s+(\d{4})$')

tbd_records = []
bad_format = []
before_april_2026 = []
outside_window = []
month_counts = defaultdict(int)

for idx, rec in enumerate(data):
    d = rec.get('Meeting_or_Event_Date', '')
    if is_empty(d):
        tbd_records.append({'idx': idx, 'org': rec_name(rec)})
        continue
    d = d.strip()
    m = DATE_RE.match(d)
    if not m:
        bad_format.append({'idx': idx, 'org': rec_name(rec), 'value': d})
        continue
    month_str, day_str, year_str = m.group(1), m.group(2), m.group(3)
    month_name_lower = month_str.lower()
    month_names_lower = [mn.lower() for mn in MONTH_NAMES]
    if month_name_lower not in month_names_lower:
        bad_format.append({'idx': idx, 'org': rec_name(rec), 'value': d, 'reason': 'unknown month'})
        continue
    month_idx = month_names_lower.index(month_name_lower)  # 0-based
    month_num = month_idx + 1  # 1-based
    year = int(year_str)

    key = f'{MONTH_NAMES[month_idx]} {year}'
    month_counts[key] += 1

    if year < 2026 or (year == 2026 and month_num < 4):
        before_april_2026.append({'idx': idx, 'org': rec_name(rec), 'value': d})

    if year != 2026 or month_num < 4 or month_num > 8:
        outside_window.append({'idx': idx, 'org': rec_name(rec), 'value': d})

# Bad format
if not bad_format:
    print('✓ All dated records match "Month D, YYYY" format.\n')
else:
    print(f'Bad date format — {len(bad_format)} record(s):')
    for r in bad_format:
        print(f'  [{r["idx"]}] {r["org"]} → "{r["value"]}"')
    print()

# Before April 2026
if not before_april_2026:
    print('✓ No dates before April 2026.\n')
else:
    print(f'Dates BEFORE April 2026 — {len(before_april_2026)} record(s):')
    for r in before_april_2026:
        print(f'  [{r["idx"]}] {r["org"]} → "{r["value"]}"')
    print()

# Outside window
if not outside_window:
    print('✓ All dates fall within April–August 2026 window.\n')
else:
    print(f'Dates OUTSIDE April–August 2026 window — {len(outside_window)} record(s):')
    for r in outside_window:
        print(f'  [{r["idx"]}] {r["org"]} → "{r["value"]}"')
    print()

# Month distribution — sort by year then month
print('Month distribution (records with a date):')
def sort_key(k):
    parts = k.rsplit(' ', 1)
    mn, yr = parts[0], int(parts[1])
    mi = [m.lower() for m in MONTH_NAMES].index(mn.lower()) if mn.lower() in [m.lower() for m in MONTH_NAMES] else 99
    return (yr, mi)
for k in sorted(month_counts.keys(), key=sort_key):
    print(f'  {k:<20} {month_counts[k]}')
print()

# TBD
print(f'TBD records (no date) — {len(tbd_records)} total:')
for r in tbd_records:
    print(f'  [{r["idx"]}] {r["org"]}')
print()

# ════════════════════════════════════════════════════════════════════════════
# CHECK 5 — County / Region Validation
# ════════════════════════════════════════════════════════════════════════════
print(SEP)
print('CHECK 5 — COUNTY / REGION VALIDATION')
print(SEP)

OFFICIAL_COUNTIES = {
    'Alachua','Baker','Bay','Bradford','Brevard','Broward','Calhoun','Charlotte',
    'Citrus','Clay','Collier','Columbia','DeSoto','Dixie','Duval','Escambia',
    'Flagler','Franklin','Gadsden','Gilchrist','Glades','Gulf','Hamilton','Hardee',
    'Hendry','Hernando','Highlands','Hillsborough','Holmes','Indian River','Jackson',
    'Jefferson','Lafayette','Lake','Lee','Leon','Levy','Liberty','Madison','Manatee',
    'Marion','Martin','Miami-Dade','Monroe','Nassau','Okaloosa','Okeechobee','Orange',
    'Osceola','Palm Beach','Pasco','Pinellas','Polk','Putnam','Santa Rosa','Sarasota',
    'Seminole','St. Johns','St. Lucie','Sumter','Suwannee','Taylor','Union','Volusia',
    'Wakulla','Walton','Washington'
}

KNOWN_MAPPINGS = {
    'Miami-Dade':   'South Florida',
    'Duval':        'Northeast Florida',
    'Hillsborough': 'Central Florida',
    'Escambia':     'Northwest Florida',
    'Leon':         'North Central Florida',
    'Orange':       'Central Florida',
    'Polk':         'Central Florida',
    'Palm Beach':   'South Florida',
    'Broward':      'South Florida',
    'Lee':          'Southwest Florida',
    'Collier':      'Southwest Florida',
    'Volusia':      'Northeast Florida',
    'Brevard':      'Central Florida',
    'Pinellas':     'Central Florida',
}

county_region_map = defaultdict(set)   # county → set of regions
region_county_map = defaultdict(set)   # region → set of counties

for rec in data:
    county = rec.get('County', '')
    region = rec.get('Region', '')
    if not is_empty(county):
        county_region_map[county].add(region if not is_empty(region) else '(blank)')
    if not is_empty(region):
        region_county_map[region].add(county if not is_empty(county) else '(blank)')

unique_counties = sorted(county_region_map.keys())
print(f'Unique counties in data ({len(unique_counties)}):')
print('  ' + ', '.join(unique_counties) + '\n')

# Invalid counties
invalid = [c for c in unique_counties if c not in OFFICIAL_COUNTIES]
if not invalid:
    print('✓ All county names match the official Florida county list.\n')
else:
    print(f'Invalid/unrecognised counties — {len(invalid)}:')
    for c in invalid:
        print(f'  "{c}" → found in regions: {", ".join(sorted(county_region_map[c]))}')
    print()

# Missing counties
missing = sorted(c for c in OFFICIAL_COUNTIES if c not in county_region_map)
print(f'Counties in official list but NOT in data ({len(missing)}):')
if missing:
    print('  ' + ', '.join(missing))
print()

# Verify known mappings
print('Verifying known County → Region mappings:')
mapping_errors = 0
for county, expected_region in KNOWN_MAPPINGS.items():
    actual_regions = county_region_map.get(county)
    if not actual_regions:
        print(f'  [NOT IN DATA] {county} — expected: {expected_region}')
    else:
        regions = sorted(actual_regions)
        if all(r == expected_region for r in regions):
            print(f'  ✓ {county:<16} → {expected_region}')
        else:
            print(f'  ✗ {county:<16} → expected "{expected_region}", found: {", ".join(regions)}')
            mapping_errors += 1

if mapping_errors == 0:
    print('\n✓ All checked mappings are correct.\n')
else:
    print(f'\n{mapping_errors} mapping error(s) found.\n')

# Complete region→county listing
print('Complete Region → County mapping in data:')
for region in sorted(region_county_map.keys()):
    counties = sorted(region_county_map[region])
    print(f'\n  {region} ({len(counties)} counties):')
    print(f'    {", ".join(counties)}')
print()

# ════════════════════════════════════════════════════════════════════════════
# CHECK 7 — "Unverified" Label Check
# ════════════════════════════════════════════════════════════════════════════
print(SEP)
print('CHECK 7 — "UNVERIFIED" LABEL CHECK')
print(SEP)

# Literal "Unverified" in any field
literal_unverified = []
for idx, rec in enumerate(data):
    hits = [k for k, v in rec.items() if isinstance(v, str) and 'Unverified' in v]
    if hits:
        literal_unverified.append({'idx': idx, 'org': rec_name(rec), 'fields': hits})

if not literal_unverified:
    print('✓ No records contain the literal string "Unverified".\n')
else:
    print(f'Records containing literal "Unverified" — {len(literal_unverified)}:')
    for r in literal_unverified:
        print(f'  [{r["idx"]}] {r["org"]} (in fields: {", ".join(r["fields"])})')
    print()

# Empty Contact_Name, Website, Source_URL (render as "Unverified" in UI)
empty_contact = [(i, r) for i, r in enumerate(data) if is_empty(r.get('Contact_Name'))]
empty_website = [(i, r) for i, r in enumerate(data) if is_empty(r.get('Website'))]
empty_source  = [(i, r) for i, r in enumerate(data) if is_empty(r.get('Source_URL'))]

print('Empty field counts (render as "Unverified" in UI):')
print(f'  Contact_Name  : {len(empty_contact)}')
print(f'  Website       : {len(empty_website)}')
print(f'  Source_URL    : {len(empty_source)}')
print()

if empty_contact:
    print(f'  Contact_Name blanks — {len(empty_contact)} records (goal: ZERO):')
    for idx, rec in empty_contact:
        print(f'    [{idx}] {rec_name(rec)}')
    print()
else:
    print('  ✓ Contact_Name: ZERO blanks — goal met.\n')

if empty_website:
    print(f'  Website blanks — {len(empty_website)} records:')
    for idx, rec in empty_website:
        print(f'    [{idx}] {rec_name(rec)}')
    print()

if empty_source:
    print(f'  Source_URL blanks — {len(empty_source)} records:')
    for idx, rec in empty_source:
        print(f'    [{idx}] {rec_name(rec)}')
    print()

print(SEP)
print('QC COMPLETE')
print(SEP)
